### PR TITLE
Add the ability to register additional controllers

### LINKED ISSUE #554: Version is not shown with self-bootstrapped controller
When viewing the controllers page with Juju 2.8 the version column on the controller page is empty 

Screen Shot 2020-06-15 at 3 49 37 PM

### LINKED ISSUE #565: Update controller page pie chart design
https://app.zeplin.io/project/5d2dfd8e5b67f14c5e1d2aef/screen/5defb971bac4ceb0d5bca38d

### LINKED ISSUE #589: Erroneous console error appearing on page load
Screenshot 2020-07-07 at 13 58 22

### PR DESCRIPTION
**Pull request by hatched**
## Done

Adds the ability to register additional controllers on the controllers page and have that controllers models added to the model list page.

## QA

Bootstrap two separate controllers on your provider of choice...
- `juju bootstrap aws test1`  
- `juju bootstrap aws test2`  

### JAAS
- Setup the dashboard for JAAS using [this guide](https://github.com/canonical-web-and-design/jaas-dashboard#in-jaas)
- Ensure that everything works as expected for the usual JAAS experience.
- Add your controllers
  - Switch to the Controllers page and click the "Register a Controller" button.
  - Follow the steps in the application.
  - Refresh
- Your `default` and `controller` models for both controllers should now be in the models list
- The two controllers should now be shown in the "Registered Controllers" section.

### Juju
- Build the tarball and update your `test1` controller from this branch with [this guide](https://github.com/canonical-web-and-design/jaas-dashboard#uploading-the-tarball-to-a-juju-controller)
- Add the `test2` controller and the `jaas` controller using the steps above.
- Your `default`, `controller`, and JAAS models for both controllers should now be in the models list.
- The two controllers should now be shown in the "Registered Controllers" section.

## Details

Still to do:
- [x] Update tests

For a follow-up:
- [ ] Do not require refresh after adding a controller
- [x] UI Enhancements.
- [ ] Removal of controllers from the list.
- [ ] Display controller-level errors in the controller page.
- [ ] Address remaining XXX comments.
- [ ] Show all registered controllers, even if not connected.
- [ ] JAAS will be adding a 'default' controller so that one shows up in the list. For now this will be blank until JAAS is updated with this feature.
- [x] Fix counts in charts.
- [ ] Validate form input for controller registration form.
- [ ] Add the `isJuju` check on a controller level basis.
- [ ] Ask users if they would like to clear saved controllers on log-out as the credentials are stored locally.
- [ ] How do we indicate to the user that the localstorage data for controllers is bad and needs to be cleared? [comment](https://github.com/canonical-web-and-design/jaas-dashboard/pull/594/files#r457521758)
- [ ] Use the submitted controller name instead of the controller `path` in the UI.
- [ ] Implement the [accept cert flow](https://github.com/canonical-web-and-design/jaas-dashboard/pull/594#issuecomment-662511683)

Fixes #554
Fixes #589 
Fixes #565 

## Screenshots

Screen Shot 2020-07-16 at 2 47 28 PM

Screen Shot 2020-07-22 at 12 39 45 PM

### CODE REVIEW

File: src/app/check-auth.js [hunk 1]
```diff
@@ -23,27 +24,56 @@ const actionWhitelist = [
...
+/**
+  Redux middleware to enable gating actions on the respective controller
+  authentication.
+  @param {Object} action The typical Redux action or thunk to execute
+  @param {Object} options Any options that this checker needs to perform an
+    appropriate auth check.
+      wsControllerURL: The full controller websocket url that the controller
+        is stored under in redux in order to determine it's logged in status.
+*/
+export default ({ getState }) => (next) => async (action, options) => {
```
Comment (hatched) [inline src/app/check-auth.js:52]: This change adds an additional argument to the `dispatch` method. It's call signature now requires the `wsControllerURL` to perform actions that involve a specific controller.
```js
dispatch(action, options)
```

File: src/components/SlidePanel/SlidePanel.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import React from "react";
+
+import "./_slide-panel.scss";
+
+function SlidePanel({ children, onClose }) {
```
Comment (hatched) [inline src/components/SlidePanel/SlidePanel.js:5]: This is just a basic pass at this panel, we'll want to enhance its functionality in follow-ups.

File: src/components/SlidePanel/_slide-panel.scss [hunk 3]
```diff
@@ -0,0 +1,15 @@
+.slide-panel {
+  background-color: white;
+  box-shadow: 0 1px 5px 1px rgba(51, 51, 51, 0.2);
+  height: calc(100% - 25px);
+  position: absolute;
+  right: 0;
+  top: 24px; // height of the top bar
```
Comment (hatched) [inline src/components/SlidePanel/_slide-panel.scss:7]: This is flaky because every time we update the height of the top bar this will have to be changed... maybe there is a better approach to this so that it works dynamically across the application?

File: src/index.js [hunk 4]
```diff
@@ -40,16 +38,19 @@ if (process.env.NODE_ENV === "production") {
 // a few cycles before trying again.
 let checkCounter = 0;
 const checkConfigExists = () => {
-  if (!window.jujuDashboardConfig && checkCounter < 5) {
-    checkCounter++;
-    setTimeout(checkConfigExists, 500);
-    return;
+  if (!window.jujuDashboardConfig) {
+    if (checkCounter < 5) {
```
Comment (hatched) [inline src/index.js:42]: This is a driveby fix to fix #589

File: src/app/model-poller.js [hunk 5]
```diff
@@ -21,46 +21,72 @@ import {
...
-      await fetchAllModelStatuses(conn, reduxStore);
-      // Wait 30s then start again.
-      await new Promise((resolve) => {
-        setTimeout(() => {
-          resolve(true);
-        }, 30000);
-      });
-    } while (isLoggedIn(reduxStore.getState()));
+
+      if (process.env.NODE_ENV === "production") {
```
Comment (barrymcgee) [inline src/app/model-poller.js:54]: We do this check before initing Sentry at root level.. is it also needed here? If so, what can we do at the conditional at root level to ensure calling `Sentry` just gracefully fails if not on production?

File: src/app/model-poller.js [hunk 6]
```diff
@@ -21,46 +21,72 @@ import {
...
+      if (userIsControllerAdmin(conn)) {
+        fetchControllerList(
+          controllerData[0],
+          conn,
+          controllerData[4],
+          reduxStore
+        );
+        // XXX the isJuju Check needs to be done on a per-controller basis
+        if (!isJuju) {
+          // This call will be a noop if the user isn't an administrator
```
Comment (barrymcgee) [inline src/app/model-poller.js:72]: a noop?

File: src/index.js (hunk 4 above)
Comment (barrymcgee) [reply src/index.js:42]: Can we cherrypick this out into a separate branch and get it landed?

**Comment by spencerbygraves** [comment]
@hatched thanks, a couple of things...

- The header bar is too high, which is the same on the live site. Should that be a separate issue?
 
image

- We have two "model status across controllers" titles, can you remove the second one please
- @clagom should the register button be below the table header?
- The button should be sentence case, no upper case "C"
- Did you use `p-icon--help` from [here](https://vanillaframework.io/docs/patterns/icons)? The colour is wrong

Thanks

**Comment by clagom** [comment]
I think the button could be above the header, yes. To be honest I'm still not 100% convinced of its position, but it's a start. @spencerbygraves @hatched

**Comment by hatched** [comment]
@spencerbygraves I've already planned for there to be a follow-up to address the UI issues as this branch is sufficiently large enough.
 
>     * The header bar is too high, which is the same on the live site. Should that be a separate issue?

This is how it is in the master branch. If you'd like to change it feel free to file an issue and we can address it separately.

>     * We have two "model status across controllers" titles, can you remove the second one please

Oops, yep consider it gone

>     * @clagom should the register button be below the table header?

I'll leave the button where it is and we can move it once the design is ironed out.  

>     * The button should be sentence case, no upper case "C"

Done.

>     * Did you use `p-icon--help` from [here](https://vanillaframework.io/docs/patterns/icons)? The colour is wrong

I did, but it's way to dark (see email thread) So I chose the lighter of the two colour options.

File: src/app/model-poller.js (hunk 6 above)
Comment (hatched) [reply src/app/model-poller.js:72]: It means "no operation". Something is a noop if it doesn't do anything.

File: src/app/model-poller.js (hunk 5 above)
Comment (hatched) [reply src/app/model-poller.js:54]: We need to connect to the controller first to be able to get its version so it has to be done here.

We could probably create a wrapper around sentry which would have this check built in. But that would be for a follow-up, feel free to file an issue for that.

**Comment by hatched** [comment]
@spencerbygraves  I went ahead and updated some of the UI elements here....any further changes can wait for a follow-up.

>     * The header bar is too high, which is the same on the live site. Should that be a separate issue?

Fixed

>     * @clagom should the register button be below the table header?

I moved it into the header as in one of the designs, it looked better there.

I also updated the charts, see the screenshot in the top post.

**Comment by spencerbygraves** [comment]
>I did, but it's way to dark (see email thread) So I chose the lighter of the two colour options.

I didn't see an email thread about icon colours. There's two variants, light and dark theme. It should look like this:

image

File: src/app/actions.js [hunk 7]
```diff
@@ -139,6 +160,19 @@ export function logOut(getState) {
...
+      const data = window.localStorage.getItem("additionalControllers");
+      if (data) {
+        additionalControllers = JSON.parse(data);
+        additionalControllers.forEach((controller) => {
+          dispatch(storeUserPass(controller[0], controller[1]));
+        });
+      }
+    } catch (e) {
+      // XXX Add to Sentry.
+      console.log("oops bad data in the additionalControllers localStorage");
```
Comment (anthonydillon) [inline src/app/actions.js:174]: Might not just be bad data, maybe something like "Error retrieving additional controllers". Then is it worth clearing the localstorage to regain a working UI without the user having the clear the storage themselves?

File: src/components/SlidePanel/SlidePanel.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  return (
+    <div className="slide-panel">
+      <div className="slide-panel--close">
```
Comment (anthonydillon) [inline src/components/SlidePanel/SlidePanel.js:8]: `slide-panel__close` this is an element of the slide-panel not a modifier to it.

File: src/components/SlidePanel/SlidePanel.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        <span
+          className="p-icon--close"
+          onClick={onClose}
+          onKeyPress={onClose}
+          role="button"
+          tabIndex="0"
+        ></span>
+      </div>
+      <div className="slide-panel--content">{children}</div>
```
Comment (anthonydillon) [inline src/components/SlidePanel/SlidePanel.js:17]: `slide-panel__content`

File: src/components/SlidePanel/_slide-panel.scss [hunk 10]
```diff
@@ -0,0 +1,15 @@
+.slide-panel {
+  background-color: white;
```
Comment (anthonydillon) [inline src/components/SlidePanel/_slide-panel.scss:2]: background-color: $color-x-light;

File: src/components/SlidePanel/_slide-panel.scss [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  box-shadow: 0 1px 5px 1px rgba(51, 51, 51, 0.2);
```
Comment (anthonydillon) [inline src/components/SlidePanel/_slide-panel.scss:3]: `box-shadow: $box-shadow`

File: src/components/SlidePanel/_slide-panel.scss [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  height: calc(100% - 25px);
+  position: absolute;
+  right: 0;
+  top: 40px; // height of the top bar
```
Comment (anthonydillon) [inline src/components/SlidePanel/_slide-panel.scss:7]: If your going to start 40px down you need to minus that from the height. Therefore the height value above should be `calc(100% - 40px)`. As they are the same you could use a variable to keep them aligned and easier read. Removing the need for the comment.

File: src/components/SlidePanel/SlidePanel.js (hunk 8 above)
Comment (anthonydillon) [inline src/components/SlidePanel/SlidePanel.js:8]: Please use a button similar to the model pattern:
```html
<button class="p-modal__close" aria-label="Close active modal" aria-controls="modal">Close</button>
```
Then you want need to worry about `role`, `tabIndex` or the cursor styling.

File: src/juju/actions.js [hunk 13]
```diff
@@ -109,9 +110,9 @@ export function fetchModelStatus(modelUUID) {
   the supplied model info call.
   @param {String} modelInfo The response from a modelInfo call.
 */
-export function addControllerCloudRegion(modelInfo) {
+export function addControllerCloudRegion(wsControllerURL, modelInfo) {
   return async function addControllerCloudRegion(dispatch, getState) {
-    const controllers = getState()?.juju?.controllers;
+    const controllers = getState()?.juju?.controllers[wsControllerURL];
     const model = modelInfo.results[0].result;
     const updatedControllers = cloneDeep(controllers).map((controller) => {
```
Comment (anthonydillon) [inline src/juju/actions.js:117]: You might want to check controllers is not null now your referencing an item in the array

File: src/pages/Controllers/Controllers.js [hunk 14]
```diff
@@ -72,54 +77,320 @@ function Details() {
...
+    delete controllerMap[uuid];
+    return row;
+  });
+
+  const rows = controllerMap && Object.values(controllerMap).map(generateRow);
 
   return (
     <>
+      <div className="controllers--header">
+        <div className="controllers--heading">
```
Comment (anthonydillon) [inline src/pages/Controllers/Controllers.js:122]: These should be `__` to be BEM

File: src/pages/Controllers/_controllers.scss [hunk 15]
```diff
@@ -1,9 +1,63 @@
+@import "vanilla-framework/scss/vanilla";
+
 .controllers {
+  th {
+    padding-bottom: 6px;
+    padding-top: 6px;
```
Comment (anthonydillon) [inline src/pages/Controllers/_controllers.scss:6]: Why?

File: src/app/actions.js (hunk 7 above)
Comment (hatched) [reply src/app/actions.js:174]: This is an interesting problem...If there is bad data they would have to go in and manually edit/clear it out which isn't a great UX. However I don't want to just clear it as then all evidence of what caused the issue would be erased. 

I'll update the error message here and add another todo around surfacing this to the user somehow so that they can take appropriate action.

File: src/components/SlidePanel/SlidePanel.js (hunk 9 above)
Comment (hatched) [reply src/components/SlidePanel/SlidePanel.js:17]: done

File: src/components/SlidePanel/SlidePanel.js (hunk 8 above)
Comment (hatched) [reply src/components/SlidePanel/SlidePanel.js:8]: removed this in favour of your other comment.

Comment (hatched) [reply src/components/SlidePanel/SlidePanel.js:8]: Good call, done

File: src/components/SlidePanel/_slide-panel.scss (hunk 10 above)
Comment (hatched) [reply src/components/SlidePanel/_slide-panel.scss:2]: done

File: src/components/SlidePanel/_slide-panel.scss (hunk 11 above)
Comment (hatched) [reply src/components/SlidePanel/_slide-panel.scss:3]: done

File: src/components/SlidePanel/_slide-panel.scss (hunk 12 above)
Comment (hatched) [reply src/components/SlidePanel/_slide-panel.scss:7]: Done

File: src/juju/actions.js (hunk 13 above)
Comment (hatched) [reply src/juju/actions.js:117]: Good call, done.

File: src/pages/Controllers/_controllers.scss (hunk 15 above)
Comment (hatched) [reply src/pages/Controllers/_controllers.scss:6]: The header styles by default have too much padding on the top so this makes the spacing consistent between the charts and two tables.

File: src/pages/Controllers/Controllers.js (hunk 14 above)
Comment (hatched) [reply src/pages/Controllers/Controllers.js:122]: done

**Comment by hatched** [comment]
Thanks for the reviews @anthonydillon @barrymcgee The requested changes have been made and it's ready for another look.

**Comment by spencerbygraves** [comment]
@hatched thanks, can we change a few things please, I've added a design below.

- Align the title and form labels
- Align close icon to fields
- Move green buttons
- Add hr above green buttons
- The check box text should be normal not small

image

**Comment by anthonydillon** [comment]
QA points captured in a Meet.

**Comment by spencerbygraves** [comment]
@hatched I've updated the design:

image

**Comment by clagom** [comment]
> @hatched I've updated the design:
> 
> image

Are these 2 steps? Not a blocker for v1, but it just makes me wonder what I need to do as a user :D

**Comment by hatched** [comment]
Thanks for the designs @spencerbygraves @clagom, I took a few liberties and put the updated image in the description of this PR.

**Comment by spencerbygraves** [comment]
@hatched @clagom I don't mind combining the title and text for the checkboxes, but I have one question.

Originally we had the text "Visit the controller to accept..." before the checkbox, now it's the other way around. Is it dependant on accepting the certificate first?

**Comment by clagom** [comment]
> @hatched @clagom I don't mind combining the title and text for the checkboxes, but I have one question.
> 
> Originally we had the text "Visit the controller to accept..." before the checkbox, now it's the other way around. Is it dependant on accepting the certificate first?

Yes that's also what I was thinking. It's unclear to me what the order is, should I go and accept the certificate and then tick the checkbox? The wording doesn't probably point out **first** that you need to go and accept before registering.

**Comment by spencerbygraves** [comment]
@clagom @hatched could this work? I also updated the wording slightly.

image

**Comment by clagom** [comment]
Thanks @spencerbygraves I think this is clearer. What do you think about adding an 'attention' kind of icon? too much?
Screenshot 2020-07-22 at 10 56 12

**Comment by spencerbygraves** [comment]
@clagom looks good

**Comment by anthonydillon** [comment]
I think we should show the address so the user knows where they are going. We can also do something smart with `:visited`

**Comment by anthonydillon** [comment]
Also, can we disable the button until we have a validate form?/

**Comment by clagom** [comment]
Both good points

**Comment by anthonydillon** [comment]
Not for this round by it might be nice to show this as a authentication type of UI.

For example:
IMG_20200722_160844

**Comment by hatched** [comment]
All requests have been addressed, it's ready for another review/qa

File: src/pages/Controllers/_controllers.scss [hunk 16]
```diff
@@ -1,9 +1,93 @@
...
+
+    .controller-link-message {
+      padding-left: 2rem;
+    }
+  }
+
+  .controllers--registered-tooltip {
+    @include vf-icon-help($color-mid-light);
+
+    cursor: pointer;
```
Comment (anthonydillon) [inline src/pages/Controllers/_controllers.scss:90]: Can you change this it `help` I clicked it and expected it to do something.

File: src/pages/Controllers/_controllers.scss [hunk 17]
```diff
@@ -1,9 +1,93 @@
...
+    }
+  }
+
+  .register-a-controller {
+    margin-top: 1rem;
+    text-align: right;
+  }
+
+  .slide-panel__content {
+    width: 550px;
```
Comment (anthonydillon) [inline src/pages/Controllers/_controllers.scss:54]: Can you change this to `max-width` (for small screens)

File: src/components/SlidePanel/_slide-panel.scss [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+.slide-panel {
+  $top-header-height: 40px;
+
+  background-color: $color-x-light;
+  box-shadow: $box-shadow;
+  height: calc(100% - #{$top-header-height});
+  overflow: scroll;
+  position: fixed;
+  right: 0;
+  top: $top-header-height;
```
Comment (anthonydillon) [inline src/components/SlidePanel/_slide-panel.scss:12]: Can you add margin-left that matches the width of the collapsed navigation panel so this UI works nicely on small screen.

File: src/pages/Controllers/_controllers.scss (hunk 16 above)
Comment (hatched) [reply src/pages/Controllers/_controllers.scss:90]: Good call, done

File: src/pages/Controllers/_controllers.scss (hunk 17 above)
Comment (hatched) [reply src/pages/Controllers/_controllers.scss:54]: done

File: src/components/SlidePanel/_slide-panel.scss (hunk 18 above)
Comment (hatched) [reply src/components/SlidePanel/_slide-panel.scss:12]: done

**Comment by hatched** [comment]
Thanks everyone for all the feedback on this PR!

:shipit:
